util/clusterauth/clusterauth.go

```diff
@@ -212,42 +216,119 @@ func InstallClusterManagerRBAC(clientset kubernetes.Interface, ns string, namesp
+	if err != nil {
+		return "", err
+	}
+
 	var secret *corev1.Secret
-	var err error
-	err = wait.Poll(500*time.Millisecond, 30*time.Second, func() (bool, error) {
-		serviceAccount, err = clientset.CoreV1().ServiceAccounts(ns).Get(context.Background(), sa, metav1.GetOptions{})
+	err = wait.PollImmediate(500*time.Millisecond, timeout, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), ClientGoRequestTimeout)
+		defer cancel()
+		secret, err = clientset.CoreV1().Secrets(ns).Get(ctx, secretName, metav1.GetOptions{})
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to get secret %q for serviceaccount %q: %w", secretName, sa, err)
 		}
-		// Scan all secrets looking for one of the correct type:
-		for _, oRef := range serviceAccount.Secrets {
-			var getErr error
-			secret, err = clientset.CoreV1().Secrets(ns).Get(context.Background(), oRef.Name, metav1.GetOptions{})
-			if err != nil {
-				return false, fmt.Errorf("Failed to retrieve secret %q: %v", oRef.Name, getErr)
-			}
-			if secret.Type == corev1.SecretTypeServiceAccountToken {
-				return true, nil
-			}
+
+		_, ok := secret.Data["token"]
+		if !ok {
+			return false, nil
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get token for serviceaccount %q: %w", sa, err)
+	}
+
+	return string(secret.Data["token"]), nil
+}
+
+// getOrCreateServiceAccountTokenSecret will check if a ServiceAccount
+// already has a kubernetes.io/service-account-token secret associated
+// with it or creates one if the ServiceAccount doesn't have one. This
+// was added to help add k8s v1.24+ clusters.
+func getOrCreateServiceAccountTokenSecret(clientset kubernetes.Interface, sa, ns string) (string, error) {
+	// Wait for sa to have secret, but don't wait too
+	// long for 1.24+ clusters
+	var serviceAccount *corev1.ServiceAccount
+	err := wait.PollImmediate(500*time.Millisecond, 5*time.Second, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), ClientGoRequestTimeout)
+		defer cancel()
+		var getErr error
+		serviceAccount, getErr = clientset.CoreV1().ServiceAccounts(ns).Get(ctx, sa, metav1.GetOptions{})
+		if getErr != nil {
+			return false, fmt.Errorf("failed to get serviceaccount %q: %w", sa, getErr)
 		}
-		return false, nil
+		if len(serviceAccount.Secrets) == 0 {
+			return false, nil
+		}
+		return true, nil
 	})
+	if serviceAccount == nil {
+		if err != nil {
+			return "", fmt.Errorf("failed to create serviceaccount token secret: %w", err)
+		}
+		log.Errorf("Unexpected nil serviceaccount '%s/%s' with no error returned", ns, sa)
+		return "", fmt.Errorf("failed to create serviceaccount token secret: nil serviceaccount returned for '%s/%s' with no error", ns, sa)
+	}
+
+	outerCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	for _, s := range serviceAccount.Secrets {
```

Comment:
According to one [article](https://itnext.io/big-change-in-k8s-1-24-about-serviceaccounts-and-their-secrets-4b909a4af4e0), the `Secrets` field won't be populated with the manually-created secret reference. So we should try to get the secret by annotation before creating a new one.

Should we add an Argo CD-internal annotation so it's easier to identify the Secret we created?

Reply:
Yep, that makes sense to me! Logic's starting to get a little complex, but seems necessary.

Reply:
Actually, I am not sure if the custom annotation is needed. Unfortunately, you can't query by annotations, just labels. So we would need to pull back all secrets regardless. FieldSelector also does not work with secret type, so it might just be that we would need to iterate over all secrets in namespace to find the correct type/annotation.

Reply:
I think we can simplify a bit and make it a bit more performatic. 
I'm thinking about having the logic such as:
1. Retrieve the provided service account (SA)
2. Inspect for existing secrets linked in the SA
2.1. If there are no secrets create a SA type secret and assign it to the SA
2.2. If there are secrets but none are of type SA token create create a SA type secret and assign it to the SA
3. return the SA assigned token

This way we don't need to retrieve all secrets in the namespace nor have to define specific annotations to search for. Do you think this simplified version would work or I am missing something?

Reply:
I misunderstood. By "assign it to the SA," you mean patch the ServiceAccount `secrets` field to add the secret we created.

Reply:
Correct. My suggestion is basically automating the steps that are now removed from 1.24 as a short term solution.
The longer term solution is approach #2 (in the PR description) leveraging the TokenRequest API.

@danielhelfand pls let me know if you are in sync with the suggestion and agree with it. I'm happy to jump on a call to clarify things if necessary.

Reply:
So recapping here:
* @leoluz and I reviewed, and think the current implementation is satisfying the cases mentioned above
* The main impact of this will be helping argocd cli users add clusters via `argocd cluster add CONTEXTNAME`
* The longer term goal here will be to migrate to the TokenRequest API but is out of scope for this pr
* Using annotations would be difficult in this case, and it is probably reasonable to ask users to add tokens for their serviceaccount in advance. Worst case is this will result in a duplicate token secret as we currently are always checking in advance if the serviceaccount has secrets.